hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -0,0 +1,460 @@
+package is.hail.expr.ir
+
+import is.hail.expr.types.physical._
+import is.hail.expr.types.virtual.{TArray, TTuple}
+import is.hail.utils._
+
+// Env contains all of the symbols values to reference
+// at root, the env is empty
+// In body of say tablemaprows, you are allowed to reference row, or global, etc
+// the names of row and global are bound in the body of tablemaprows
+// so tablemaprows appends to the environment
+// It's possible for a name to be bound twice
+// we may have nested array maps, and say each binds i
+// inside the inner map body, any reference to the name
+// FIXME: strip all requiredness logic when possible
+// TODO: We don't implicitly walk bottom-up for virtualTypes (not calling .typ) in all cases
+// TODO: If virtual type inference doesn't occur before this step, there may be subtle
+// TODO: behavior differences between IR Nodes (those that need .typ here and those that don't)
+// TODO: In cases where env is sent as is to child nodes, do we need copy? I believe no, as long as copy-on-write
+
+// TODO: aggregators: 1) use bindAgg/bindScan instead of bindEval? 2) proper passing of env, seems little needed
+object InferPType {
+  def apply(ir: IR, env: BindingEnv[PType]): PType = {
+    ir match {
```

Comment:
you can add a helper method above here:
```scala
def infer(x: IR, env: Env[PType] = env): Unit = InferPType(x, env)
```

Then use that below to make things a little cleaner